core/DataAccess/LogAggregator.php

```diff
@@ -373,35 +383,75 @@ public function generateQuery($select, $from, $where, $groupBy, $orderBy, $limit
                         break;
                     }
                 }
-
             }
 
         }
 
         $query = $segment->getSelectQuery($select, $from, $where, $bind, $orderBy, $groupBy, $limit, $offset);
 
+        if (is_array($query) && array_key_exists('sql', $query)) {
+            $query['sql'] = $this->addOriginHintToQuery($query['sql']);
+        }
+
+        return $query;
+    }
+
+    /**
+     * Create the segment temporary table
+     *
+     * @return string   Name of the created temporary table, including any table prefix
+     *
+     * @throws \DI\DependencyException
+     * @throws \DI\NotFoundException
+     */
+    private function createSegmentTable(): string
+    {
+        $segmentTable = $this->getSegmentTmpTableName();
+        $segmentWhere = $this->getWhereStatement('log_visit', 'visit_last_action_time');
+        $segmentBind = $this->getGeneralQueryBindParams();
+
+        $logQueryBuilder = StaticContainer::get('Piwik\DataAccess\LogQueryBuilder');
+        $forceGroupByBackup = $logQueryBuilder->getForcedInnerGroupBySubselect();
+        $logQueryBuilder->forceInnerGroupBySubselect(LogQueryBuilder::FORCE_INNER_GROUP_BY_NO_SUBSELECT);
+        $segmentSql = $this->segment->getSelectQuery('distinct log_visit.idvisit as idvisit', 'log_visit', $segmentWhere, $segmentBind, 'log_visit.idvisit ASC');
+        $logQueryBuilder->forceInnerGroupBySubselect($forceGroupByBackup);
+
+        $this->createTemporaryTable($segmentTable, $segmentSql['sql'], $segmentSql['bind']);
+
+        return $segmentTable;
+    }
+
+    /**
+     * Add an origin hint to the query to identify the main parameters and segment for debugging
+     *
+     * @param string $sql   SQL query string
+     *
+     * @return string   Modified SQL query string with hint added
+     */
+    private function addOriginHintToQuery(string $sql): string
```

Comment:
nice we're moving these things into separate methods 👍 

Something to think about could be moving those methods into a separate class where those private methods are then public meaning it allows us to write few basic tests for them.
Just a thought. 

Same could be potentially done for the other private method `createSegmentTable`

Reply:
I've made the  `addOriginHintToQuery()` method static, adjusted it to receive the hint fields as parameters and moved it to `core/DbHelper` to sit next to `addMaxExecutionTimeHintToQuery()`. All parameters are optional except the origin string, so this could potentially be useful elsewhere if we want to tag another non-segment query for debugging. I've added a new integration test to check the expected output.

`createSegmentTable` uses a number of other private methods in `LogAggregator`, some of which look like they might need adjusting as part of the database abstraction changes (schema creation syntax, explicit engine parameter, etc), so it might be better to refactor segment table creation into a separate helper class as part of those changes.